qutip/core/cy/coefficient.pyx

```diff
@@ -365,167 +373,180 @@ def coeff(t, args):
-        self.tlist_np = tlist
-        self.tlist = tlist
-        self.coeff_np = coeff_arr
-        self.coeff_arr = coeff_arr
-        if _second is None:
-            self.second_np, self.constant = _prep_cubic_spline(coeff_arr,
-                                                                tlist)
+    cdef complex[:, :] poly
+    cdef object np_arrays
+
+    def __init__(self, coeff_arr, tlist, int order):
+        tlist = np.array(tlist, dtype=np.float64)
+        coeff_arr = coeff_arr.astype(np.complex128)
+
+        if coeff_arr.ndim != 1:
+            raise ValueError("The array to interpolate must be a 1D array")
+        if coeff_arr.shape != tlist.shape:
+            raise ValueError("tlist must be the same len "
+                             "as the array to interpolate")
+        if order < 0:
+            raise ValueError("order must be a positive integer")
+
+        order = min(order, len(tlist) - 1)
+
+        if order == 0:
+            self.np_arrays = (tlist, coeff_arr.reshape((1, -1)))
+        elif order == 1:
+            self.np_arrays = (
+                tlist,
+                np.vstack([
+                    np.diff(coeff_arr, append=-1) / np.diff(tlist, append=-1),
+                    coeff_arr
+                ]))
+        elif order >= 2:
+            # Use scipy to compute the spline and transform it to polynomes
+            # as used in scipy's PPoly which is easier for us to use.
+            spline = make_interp_spline(tlist, coeff_arr, k=order)
+            # Scipy can move knots, we add them to tlist
+            ts = np.sort(np.unique(np.concatenate([spline.t, tlist])))
+            a = np.arange(spline.k+1)
+            a[0] = 1
+            fact = np.cumprod(a)
+            poly = np.concatenate([
+                spline(ts, i) / fact[i]
+                for i in range(spline.k, -1, -1)
+            ]).reshape((spline.k+1, -1))
+            self.np_arrays = (ts, poly)
+
+        self._prepare()
+
+    def _prepare(self, dt=None):
+        self.tlist = self.np_arrays[0]
+        self.poly = self.np_arrays[1]
+        self.order = self.poly.shape[0] - 1
+        diff = np.diff(self.np_arrays[0])
+        if dt is not None:
+            self.dt = dt
+        elif len(diff) >= 1 and np.allclose(diff[0], diff):
+            self.dt = diff[0]
         else:
-            self.second_np = _second
-            self.constant = _constant
-        self.second_derr = self.second_np
-        self.dt = tlist[1] - tlist[0]
+            self.dt = 0
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    @cython.cdivision(True)
+    cdef size_t _binary_search(self, double x):
+        # Binary search for the interval
+        # return the indice of the of the biggest element where t <= x
+        cdef size_t low = 0
+        cdef size_t high = self.tlist.shape[0]
+        cdef size_t middle
+        cdef size_t count = 0
+        while low+1 != high and count < 30:
```

Comment:
Why do we keep track of `count` here?

Reply:
The ensure that it does not go into an infinite loop even if input are wrong.
If one manually build the object with `restore` or a subclass he could have a `tlist` with 0 element, etc. 
Since cython does not respond to interrupt signal, I prefer being safe.

Reply:
Could we leave a comment to that effect and maybe bump the limit up a bit higher -- e.g. 64. 30 is probably high enough, but a list of, e.g., 2**32 floats is very achievable on modern machines (even if it's perhaps a bit unrealistic).